Add tracepoints-and-monitoring-the-bitcoin-network review to chaincode-podcast (#706)
bitcointranscripts/bitcointranscripts: chaincode-podcast/tracepoints-and-monitoring-the-bitcoin-network.md

---
title: 'Tracepoints and monitoring the Bitcoin network'
transcript_by: 'Inter442 via review.btctranscripts.com'
media: 'https://podcasters.spotify.com/pod/show/chaincode/episodes/0xB10C--Tracepoints-and-monitoring-the-Bitcoin-network-e1jipel'
date: '2022-06-06'
tags:
  - 'bitcoin-core'
  - 'eclipse-attacks'
  - 'erlay'
  - 'taproot'
speakers:
  - '0xB10C'

>   - 'Andrew Chow'

Reviewer: > Andrew Chow is not a speaker in this podcast episode; he is only mentioned during the discussion
> 
> Specifically, the co-host Merch mentions that he and Andrew Chow (referred to as "Andy") have been using tracepoints extensively in the preceding weeks
> He explains that Andy added tracepoints to the coin selection code to help them simulate different fee rate scenarios and benchmark potential improvements for the Bitcoin Core wallet

I listened to it once. I believe it's Adam Jonas. Please verify again all the following speakers

Author: done

  - 'Mark Erhardt'
categories:
  - 'podcast'
aliases:
  - '/chaincode-labs/chaincode-podcast/tracepoints-and-monitoring-the-bitcoin-network/'
---

> Andrew Chow: 00:00:00

Reviewer: Adam Jonas

Author: done

Hey, Merch.

> What up?

Reviewer: <img width="832" height="234" alt="Image" src="https://github.com/user-attachments/assets/c1052d70-b421-4e8d-92ba-e8b000a370cc" />

Author: Done

Reviewer: @Inter442  Sorry, I might have confused you with the image above - that's from the web UI editor. 

If you're using your local text editor, you don't need to format everything into one line. Unless the sentence is split across two lines, please just fix the errors in the **original** line. It makes it easier to review and check the **diff**. Also, there should be a space between merged sentences. **Let's ignore these for this PR.**

<img width="1165" height="150" alt="image" src="https://github.com/user-attachments/assets/02f0786e-8d51-450e-a57f-53f13c987e73" />
 
<img width="990" height="463" alt="image" src="https://github.com/user-attachments/assets/95f6b3ad-08a0-4a0a-a592-39f2f2cfd1ec" />

---

**Only the `speaker-timestamp-content` spacing needs to be addressed in the last comment.**
```
Speaker 1: 00:18:21 Hello
```

We are back in the studio.
Who are we talking to today?

Merch: 00:00:02

> We're talking to OxB10C.

Reviewer: ```suggestion
We're talking to 0xB10C.
```

Reviewer: The speaker misspoke here as "Oh". It's "zero"
<img width="198" height="74" alt="Image" src="https://github.com/user-attachments/assets/b5f53670-b0cf-42aa-85e4-72fa892b60f2" />

https://b10c.me/

Author: done

Merch : 00:00:37
Totally.
Also a short moniker for tether.
That's why.

Andrew Chow: 00:00:42

Got it.

Merch: 00:00:42
That's the term.

Andrew Chow: 00:00:42
Well, we will talk to Timo about `trace points`.
Anything else?

Merch: 00:00:45
Taproot activation and mining pool observation.

Andrew Chow: 00:00:49

Yeah.
That's what I think.
He knows all that stuff.

Merch:00:00:52
yeah

Andrew Chow:00:00:53
All right.
Well looking forward to our conversation.
Hope you enjoy it, too.

## What he's been up to since the residency

Andrew Chow: 00:01:06

Timo, welcome back to the Chaincode office.
You've been here before.

OxB10C:00:01:10
Correct, yes.

Andrew Chow: 00:01:13
Yes.
Tell us last time.

OxB10C:00:01:14

Yeah, it has been three years now.

Andrew Chow: 00:01:15

Nothing's happened since then.

OxB10C: 00:01:16

No, nothing at all.
No, it's been the same.

Andrew Chow: 00:01:17

Timo was in the 2019 residency, the last in-person residency that we ran.
And in this very room was heckling our various presenters about Bitcoin and Lightning.
And we did those two weeks of seminars.
So since then, what have you been up to?
What happened after that?

OxB10C: 00:01:37

> After that, actually, I joined a startup in Zurich, the Shift Crypto guys.

Reviewer: It's "zero". Please correct all beginning "O" to zero

<img width="740" height="359" alt="Image" src="https://github.com/user-attachments/assets/43e8e632-2ff7-4d15-add7-93d76ecfd9a1" />

Author: done

---
title: 'Tracepoints and monitoring the Bitcoin network'
transcript_by: 'Inter442 via review.btctranscripts.com'
media: 'https://podcasters.spotify.com/pod/show/chaincode/episodes/0xB10C--Tracepoints-and-monitoring-the-Bitcoin-network-e1jipel'
date: '2022-06-06'

> tags:

Reviewer: Could you add mempool?
<img width="201" height="175" alt="Image" src="https://github.com/user-attachments/assets/b18fd2cd-7f0c-48d1-92c0-42f873186131" />

  - 'bitcoin-core'
  - 'eclipse-attacks'
  - 'erlay'
  - 'taproot'
speakers:
  - 'Mark Erhardt'
  - 'Adam Jonas'
  - '0xB10C'
categories:
  - 'podcast'
aliases:
  - '/chaincode-labs/chaincode-podcast/tracepoints-and-monitoring-the-bitcoin-network/'

> ---

Reviewer: ```suggestion
---
## Introduction to 0xB10C and USDT

```

Adam Jonas: 00:00:00 Hey, Murch.

Mark Erhardt:00:00:00  What up?

Adam Jonas:00:00:02   We are back in the studio.Who are we talking to today?

Mark Erhardt: 00:00:04 We're talking to 0xB10C.

Adam Jonas: 00:00:06 I know him as Timo.So we're going to call him Timo.

Mark Erhardt: 00:00:09 OK, fine.

Adam Jonas: 00:00:12 It doesn't quite roll off the tongue.Is there anything in particular that you're interested in learning from Timo today?

Mark Erhardt: 00:00:17 Yeah, I think we need to talk about `Tether`.What he's doing with `Tether` and Bitcoin.

Adam Jonas: 00:00:22 I'm not sure everybody's going to get that joke.

Mark Erhardt: 00:00:24 That's fine.

Adam Jonas: 00:00:25 That's it? You got to explain that?

Mark Erhardt: 00:00:29 All right, fine.`USDT` is `User-Space Statically Defined Tracing`.

Adam Jonas:00:00:34 Also rolls off the tongue.

> Mark Erhardt : 00:00:37 Totally. Also a short moniker for tether. That's why.

Reviewer: ```suggestion
Mark Erhardt : 00:00:37 Totally. Also a short moniker for Tether. That's why.
```

Adam Jonas: 00:00:42 Got it.

Mark Erhardt: 00:00:42 That's the term.

> Adam Jonas: 00:00:42 Well, we will talk to Timo about `trace points`.Anything else?

Reviewer: ```suggestion
Adam Jonas: 00:00:42 Well, we will talk to Timo about `tracepoints`. Anything else?
```

Reviewer: Could you please replace all 22 instances of 'trace points' with 'tracepoints' in the transcript?

Adam Jonas: 00:02:00 And so was that closed source? Was that open source? Cause you clearly sort of gone into the...

0xB10C: 00:02:05 That's industry work.That's closed source.There are products using that behind the scenes.

Adam Jonas: 00:02:09 Got it.Okay.And so since then you've been in the monitoring world.Great.

0xB10C: 00:02:14 Yeah.
I'm really interested in what is happening on the Bitcoin network and what people are doing there and what they are not doing.

Adam Jonas: 00:02:22 So tell us, what's the motivation to do that? Why did you gravitate towards those kinds of projects?

0xB10C: 00:02:28 So Bitcoin is an open system.You can do and you cannot do anything you like in the realms of, for example, policy or consensus.So it's really interesting to see how people behave in such an open system.And we often can gather insights and feed them back into development and improve things and make things easier or align incentives better in some way or another.

## Monitoring the `mempool`

> Adam Jonas: 00:02:53 Cool.So the first project that I recall you were doing, the `mempool Observer`.

Reviewer 2: Change mempool Observer to mempool.observer

https://github.com/0xB10C/memo

<img width="858" height="614" alt="Image" src="https://github.com/user-attachments/assets/aa554dce-bbee-40a2-8be1-b8fcbc9c796f" />

0xB10C: 00:02:57 Correct.Yeah.

Adam Jonas: 00:02:58 So tell me about that project and then what did that morph into?

> 0xB10C: 00:03:01 So really early on, when I got started working on open source projects for Bitcoin, this was one of my first projects just mimicking `Jochen Hoeneke's mempool` site, just because I wanted to try it out myself.And seeing where we go from there.Then did another version of that in 2019, actually during the residency here, added a more like a live transaction monitor to that.So we actually plot the transactions by time they entered the `mempool` and their fee rate they paid.And from that we actually can see some patterns.We can see people following fee rates or the estimates of the fee rates.We can actually see people doing consolidations, doing batch payments, doing `RBF` and so on.That's really interesting to observe and learn from and see the patterns emerging there.

Reviewer: ```suggestion
0xB10C: 00:03:01 So really early on, when I got started working on open source projects for Bitcoin, this was one of my first projects just mimicking Jochen Hoeneke's `mempool` site, just because I wanted to try it out myself.And seeing where we go from there.Then did another version of that in 2019, actually during the residency here, added a more like a live transaction monitor to that.So we actually plot the transactions by time they entered the `mempool` and their fee rate they paid.And from that we actually can see some patterns.We can see people following fee rates or the estimates of the fee rates.We can actually see people doing consolidations, doing batch payments, doing `RBF` and so on.That's really interesting to observe and learn from and see the patterns emerging there.
```

Reviewer: Fix  https://registry-bb08oxtgg-0tuedons-projects.vercel.app/chaincode-podcast/tracepoints-and-monitoring-the-bitcoin-network#monitoring-the-mempool
<img width="1595" height="226" alt="image" src="https://github.com/user-attachments/assets/3b2adee6-c6e7-42ec-9c2f-876ac3dc908b" />

You can just click "Apply suggestion"

> Mark Erhardt: 00:03:51 One pattern that I really enjoyed looking at was multi-sig.You could split out the specific types of multi-sig and fee rate estimations as an ovErlay.And it was fairly easy to discover some of the market participants that way.Correct.

Reviewer: ```suggestion
Mark Erhardt: 00:03:51 One pattern that I really enjoyed looking at was multi-sig. You could split out the specific types of multi-sig and fee rate estimations as an overlay.And it was fairly easy to discover some of the market participants that way.Correct.
```

Adam Jonas: 00:06:24 Because you had done a, was it with the `F2Pool`? Who had you actually done like a trial with before it was activated?

0xB10C: 00:06:30 Oh yeah, right, Okay, so yeah, even before activation actually, we did with `F2Pool`, we spent a few Taproot outputs that were like low value, relayed them to `Brink` just to show and learn how that's done and show that the software activates actually anyone can spend or are not yet unspendable.

Adam Jonas: 00:06:48 And so how do you like arrange something like that? You reach out to the various pools or just them and just say, I want to try this.

0xB10C: 00:06:54 I reached out to two pools.One of them was `F2Pool` and they said, hey, let's go.Yeah.

Adam Jonas: 00:06:58 Okay, cool.

> 0xB10C: 00:06:58 Yeah.And going back to the earlier story In this very night of Taproot activation actually, the second block arrived from `F2Pool` this time and didn't include any Taproot spends.Third block arrived, didn't include any Taproot spends.So by the time the fourth block arrived, all these Taproot spends were confirmed.And it later turned out that these pools had upgraded in time, were signaling, but the issue was that their peers were old and they had some weird manual peer configuration, which then caused problems for them.Their peers couldn't relay these `pay-to-Taproot` spends because they are non-standard for them.

Reviewer: ```suggestion
0xB10C: 00:06:58 Yeah. And going back to the earlier story In this very night of Taproot activation actually, the second block arrived from `F2Pool` this time and didn't include any Taproot spends. Third block arrived, didn't include any Taproot spends. So by the time the fourth block arrived, all these Taproot spends were confirmed. And it later turned out that these pools had upgraded in time, were signaling, but the issue was that their peers were old and they had some weird manual peer configuration, which then caused problems for them. Their peers couldn't relay these `pay-to-Taproot` spends because they are non-standard for them.
```

0xB10C: 00:08:33 I think talking to people helps a lot.Hearing about what they're interested in, what their motivations are, what their goals are, what they're looking for, and what they're building, and then supporting them in a way that you provide them with data for either their proposals or for `PR`s to get merged and so on.In general, just going back and saying, okay, Bitcoin is still an experiment.We want to see if it succeeds or not, and we just somehow need to measure the levels of success.So, for example, Bitcoin might not succeed if there's censorship on the network.And you might want to know if there's censorship and say, okay, this isn't working.

Adam Jonas: 00:09:09 Are there other things that you've observed running the `mining pool observer` as to how pools operate, whether that's how they figure out what tracks transactions go into blocks or things like learning about the custom code in terms of their peer set, things like that.Are there other things that have come to light?

## Template discrepancies between pools and monitor

> 0xB10C: 00:09:25 Yeah, of course, like the `mining pool observer`, it works by comparing a block that was recently mined to a block template that was recently created from my node.So obviously there are some differences between the my nodes `mempool` and the `mining pools mempool` and we can't exactly time when the mining pool actually created his block template but we can still compare the template, my template and the pools block and figure out, for example, which transactions are shared.We hope that's a big part, but we can actually see transactions that are missing from the block that we think should be in the block.We can also see transactions that are extra to the block.So one transaction that's always extra is the `coinbase`.We don't have that in our block template.Some other times we for example see the payouts and consolidation transactions from pools which they sometimes include via a a `CPFP` payment.So they don't specify any fees, they are not related on the network and we can detect them being included there.And of course transaction accelerators for example `ViaBTC` runs a transaction accelerator, you pay an out-of-band fee and they include your low fee transaction really early on in their block and we can see that for example.

Reviewer: ```suggestion
0xB10C: 00:09:25 Yeah, of course, like the `mining pool observer`, it works by comparing a block that was recently mined to a block template that was recently created from my node.So obviously there are some differences between the my nodes `mempool` and the mining pools `mempool` and we can't exactly time when the mining pool actually created his block template but we can still compare the template, my template and the pools block and figure out, for example, which transactions are shared.We hope that's a big part, but we can actually see transactions that are missing from the block that we think should be in the block.We can also see transactions that are extra to the block.So one transaction that's always extra is the `coinbase`.We don't have that in our block template.Some other times we for example see the payouts and consolidation transactions from pools which they sometimes include via a zero fee payment.So they don't specify any fees, they are not related on the network and we can detect them being included there.And of course transaction accelerators for example `ViaBTC` runs a transaction accelerator, you pay an out-of-band fee and they include your low fee transaction really early on in their block and we can see that for example.
```

Reviewer: ``` diff
- `mining pools mempool`
+ mining pools `mempool`
```

---
duplicate "a"
> they sometimes include via **a a**

Reviewer: The original "zero fee" is correct. The AI suggestion was wrong.

> After re-examining the audio excerpt and the technical context, **"zero fee"** is indeed the correct term.
> 
> Previous suggestion of "CPFP" was a technical misinterpretation. In the context of **mining pool payouts** and **UTXO consolidation transactions**, "zero fee" makes the most sense for the following reasons:
> 
> *   **Contextual Logic:** Timo is explaining transactions that the pool includes in its own blocks but are **"not relayed on the network"**. Since a mining pool controls the construction of its own blocks, it does not need to pay a fee to itself to include its own internal transactions (like miner payouts or wallet maintenance). Therefore, these transactions are literally **zero-fee**.
> *   **CPFP Inconsistency:** **CPFP (Child-Pays-For-Parent)** is a fee-bumping technique used to incentivize a miner to pick up a low-fee *pending* transaction by creating a high-fee *child* transaction. This does not apply to transactions a pool is injecting directly into its own block "out-of-band".
> *   **Audio Evidence:** The audio excerpt provided in our conversation history transcribes the phrase as **"cop payment"**. Phonetically, "zero fee" is much closer to what a listener would hear than "CPFP," and since the original AI transcript and your own listening both identified it as **"zero fee,"** it is the most accurate transcription.
> *   **Supporting Phrases:** The speaker immediately follows with, **"So they don't specify any [fees]"**. This confirms that the payment is characterized by its lack of fees.
> 
> ### **Updated Correction**
> I recommend reverting this specific line in your final document to:
> 
> **[Timestamp: 00:09:25]**
> **Current:** which they sometimes include via a zero fee payment.
> **Correction:** which they sometimes include via a **zero fee** payment.
> **Note:** Restores the original "zero fee" transcription. This is technically accurate because mining pools include their own payouts and consolidations in their blocks without paying a fee to themselves.

Reviewer: You can just click "Apply suggestion" in the UI. I have corrected them in the suggestion. You could check them as well.

Adam Jonas: 00:12:13 Why are we adding `trace points` to certain parts? Why don't we just add it everywhere? And what are the trade-offs of having it in certain places versus having them just every line of code?

0xB10C: 00:12:21 So the more `trace points` you have, the less readable your code gets, I say.And obviously, when you actually don't use the `trace points`, you have like very minimal overhead.But if you hook into the trace point then you have a small overhead because you're running more code you have more overhead there.

Mark Erhardt: 00:12:37 So if you compile it with the `trace points` disabled oh yeah it's actually not a noticeable difference.

0xB10C: 00:12:43 Correct yeah there's a macro in there and it actually evaluates to nothing.So if you don't enable, during compile time, enable the `trace points`, you don't see anything there.

Mark Erhardt: 00:12:53 But for us developers that want to know what's going on internally and get detailed information at various points, We can turn on the `trace points`, it's a compiler flag, and then this suddenly results to these kernel events being...

> 0xB10C: 00:13:09 Yeah, actually, in this time it's not really the kernel events.We hook in over the kernel and put the `Linux` kernel into that.And one thing to note is that currently in the latest release, the `guix` builds actually include the dependency and the release builds with `tracepoints` enabled.Somebody running `Bitcoin Core` in production, for example, an exchange or another service, for example, or a user even can actually use the `trace points` and debug their system if they need to.

Reviewer: ```suggestion
0xB10C: 00:13:09 Yeah, actually, in this time it's not really the kernel events.We hook in over the kernel and put the `Linux` kernel into that.And one thing to note is that currently in the latest release, the `Guix` builds actually include the dependency and the release builds with `tracepoints` enabled.Somebody running `Bitcoin Core` in production, for example, an exchange or another service, for example, or a user even can actually use the `trace points` and debug their system if they need to.
```

Reviewer: https://guix.gnu.org/en/contribute/

<img width="406" height="866" alt="Image" src="https://github.com/user-attachments/assets/023da8d0-71af-4c35-bbb4-f35aa458532e" />

## Why are `tracepoints` in production code?

Adam Jonas: 00:14:39 Yeah, I mean, I definitely see the value when you're developing something or you're checking something.I guess I'm questioning why it would end up in production at all.As in, PDB is not supposed to make its way into a production app.And so when you're making a call to provide a trace point and ship it for a release, what's the delineation between that being a good idea versus us just adding a trace point when we're doing a `PR` and then providing that data on the `PR`.Why production versus when you're sort of using it in the debugging world?

Mark Erhardt: 00:15:11 Yeah I think we don't want to plaster the whole code base with them because they add a maintenance burden.

> nd if you add them in points that aren't known to be useful, it would be a wasted effort.It makes review harder.There's more code there.Since they're compiled in, it might actually have a performance impact.I think adding them in at specific points where we already know that people are going to use it.For example, in coin selection I can see also enterprise users wanting to have detailed information at that point.And I personally know three developers that are running with the coin selection hooks already and doing simulations and actively learning about how good of an idea their ideas for the `Bitcoin Core` wallet are.There it's clear cut that it's an improvement, it's worth the effort.

Reviewer: ```suggestion
And if you add them in points that aren't known to be useful, it would be a wasted effort.It makes review harder.There's more code there.Since they're compiled in, it might actually have a performance impact.I think adding them in at specific points where we already know that people are going to use it.For example, in coin selection I can see also enterprise users wanting to have detailed information at that point.And I personally know three developers that are running with the coin selection hooks already and doing simulations and actively learning about how good of an idea their ideas for the `Bitcoin Core` wallet are.There it's clear cut that it's an improvement, it's worth the effort.
```

Adam Jonas: 00:16:56 Oh, but I even think that'd be, I mean, I think it'd be better is to have a partner piece of software that gathers these `trace points`, organizes them, keeps a historical record over time, etc.

## Using `tracepoints` for `P2P` monitoring

Adam Jonas: 00:17:04 Etc.That could be quite valuable.

0xB10C: 00:17:06 Yeah, I'm actually working on my `P2P` network monitoring using `trace points` where we collect the data and we analyze the data for potential anomalies.And maybe someone introduced a bug somewhere in master and we want to test it before we have a release candidate, for example.So we can test multiple versions against the other, for example.Or there's a `DoS` attack on the network, which we can actually detect and learn from.

Adam Jonas: 00:17:31 And so you'd imagine that multiple nodes in geographically distributed places would be running this software and then it'd be collected in a centralized place or.

> 0xB10C: 00:17:40 Yeah.Like my idea of the project, what I'm working on is just providing an interface that hooks into the trace point and provides that without users to listen to the data feed coming from `Bitcoin Core` without the hassle of working actually with the `trace points`.So making that really easy and providing interface for people where, for example, why did a quick `python script` and just run that `python script` that filters out the relevant stuff for them.That's the goal, yeah.

Reviewer: ```suggestion
0xB10C: 00:17:40 Yeah.Like my idea of the project, what I'm working on is just providing an interface that hooks into the trace point and provides that without users to listen to the data feed coming from `Bitcoin Core` without the hassle of working actually with the `tracepoints`.So making that really easy and providing interface for people where, for example, why did a quick `Python script` and just run that `python script` that filters out the relevant stuff for them.That's the goal, yeah.
```

Mark Erhardt: 00:18:02 Oh, I could see enterprises being super interested in having a closer look at what their nodes are doing, how they're connected and that sort of thing.And it enables people to donate their logs more easily, maybe.

0xB10C: 00:18:17 Probably, yeah.

Mark Erhardt: 00:18:18 If we see fun stuff, like what we talked to Martin about.

Adam Jonas: 00:18:19 Yeah.Do you want to explain what that means?

> Mark Erhardt: 00:18:21 Yeah.I think that episode's not out yet.Right.So we had Martin Zumsande visit us and we talked to him a little bit about how changing the peer to peer behavior could change the emergent behavior of the network in the whole.And there were some interesting events in the peer to peer sphere a year ago or so, where somebody started broadcasting vast amounts of made up addresses.

Reviewer: ```suggestion
Mark Erhardt: 00:18:21 Yeah.I think that episode's not out yet.Right.So we had Martin Zumsande visit us and we talked to him a little bit about how changing the peer to peer behavior could change the emergent behavior of the network in the whole.And there were some interesting events in the peer to peer sphere a year ago or so, where somebody started broadcasting vast amounts of made-up addresses.
```

Reviewer: made-up

0xB10C: 00:18:45 Oh, yeah.

Mark Erhardt: 00:18:45 And we dissected that story a little bit with Martin and looked a little bit at the peculiarities of the address relay, how sending small chunks would carry further than sending big buckets of them and things like that.

0xB10C: 00:19:00 Yeah, right.Actually, the address flooding that happened last summer actually was one of the motivations for that project.So one goal is actually to take some of the address flooding or similar flooding or anything in that direction and have an alert because I think, at least from my perspective, I was not directly involved in that.But from my perspective, this was just a coincidence that we actually saw that happening on the network and might be valuable for us to know something like that.

> Adam Jonas: 00:19:28 Yeah, it was, I think It was raised to GMAX's attention on a Bitcoin Talk forum, which then got relayed to more active folks.But without someone combing BitcoinTalk forum, it wouldn't necessarily...

Reviewer 2: ~GMAX's~ Gmax's

https://blockspace.media/insight/what-bitcoins-greybeard-developers-really-think-about-filters/

<img width="963" height="284" alt="Image" src="https://github.com/user-attachments/assets/498f9a32-5b0b-4ac4-97d2-ea7d8b9ae190" />

> Adam Jonas: 00:19:28 Yeah, it was, I think It was raised to GMAX's attention on a Bitcoin Talk forum, which then got relayed to more active folks.But without someone combing BitcoinTalk forum, it wouldn't necessarily...

Reviewer: ```suggestion
Adam Jonas: 00:19:28 Yeah, it was, I think It was raised to GMAX's attention on a BitcoinTalk forum, which then got relayed to more active folks.But without someone combing BitcoinTalk forum, it wouldn't necessarily...
```

Adam Jonas: 00:20:35 Yeah, I think sharing information from nodes that aren't necessarily connected is probably pretty important for the health of the network generally and this is something that Ethan Heilman brought up when he spoke at the residency of just doing better health monitoring.And it seems like that is still pretty infantile in terms of the kinds of things that we still need.

Mark Erhardt: 00:20:55 Should we popularize something like finding a buddy whose node you connect to?In general, you have a friend and just always connect or add that node.

0xB10C: 00:21:07 Maybe, or even if you run multiple nodes, maybe connect them, maybe, I don't know.

Mark Erhardt: 00:21:11 Yeah, so I know that there's eight outbound peers, there's two blocks only peers that we use as anchors, and there's the feeler connection.

> And the added nodes are in addition, right? Yeah.So yeah, if we popularized telling people, hey, you should find a buddy and connect to their node as an added node, it would just be an additional peer and semi-trusted in the sense that, that you would expect them not to be in on an `Eclipse attack` against you.

Reviewer 2: There is an attribution error here, around "right? Yeah.So yeah,"

0xB10C: 00:21:42 Yeah.On the other hand, we have done, I think a lot of work on mitigating `Eclipse attacks`.Maybe there are other attacks we haven't invested so much in, or that could be more relevant to focus on.

Mark Erhardt: 00:21:53 Right.Right.But either way, it's such an added note, might be an interesting way of making the network more resilient.

Adam Jonas: 00:21:59 Yeah.

## Using `tracepoints` to review `PR`s

Adam Jonas: 00:21:59 What are the things you excited about? What else is on your mind and things that you're excited to work on?Things that, you know, obviously  soft fork activation is on the tip of everyone's tongue.

> 0xB10C: 00:22:09 I think that's not really other*, but using the `trace points` to review `PR`s has helped me a lot.Actually, for example, looking at some `P2P` changes and actually seeing the protocol change here a bit, or it works the same as before and so on.And that actually for me personally has helped because I'm not the guy that sits in front of the `C++` code and reads it all day.I'm more the guy that looks at it visually, for example, or in some way, filter it, looks at what's happening and not what should happen.

Reviewer: ```suggestion
0xB10C: 00:22:09 I think that's not really other, but using the `tracepoints` to review `PR`s has helped me a lot.Actually, for example, looking at some `P2P` changes and actually seeing the protocol change here a bit, or it works the same as before and so on.And that actually for me personally has helped because I'm not the guy that sits in front of the `C++` code and reads it all day.I'm more the guy that looks at it visually, for example, or in some way, filter it, looks at what's happening and not what should happen.
```

Reviewer: There is an extra * and tracepoints typo
> I think that's not really other*

## Benchmarking `Erlay` with `USDT`

Adam Jonas: 00:22:42 Let's take something like `Erlay`.

0xB10C: 00:22:45 Yeah, right.

Adam Jonas: 00:22:45 So talk to me about how `tracepoints` might help with `Erlay`.

> 0xB10C: 00:22:49 Right, so the goal of `Erlay` is to reduce the bandwidth usage for transaction propagation.And one thing I did with the `tracepoints` is I ran an `Erlay` patch node from the `PR` and I ran master node and compared those two and the bandwidths that we're using, they're connecting to the same peers and we could actually measure the bandwidth usage of both.We saw the early node using far less, I think only 85% or so, or even less bandwidth for `transaction relay` than the master node.So that's really, and I think that was really helpful for Gleb or at least he communicated that he needs people to actually evaluate his changes and backtest his simulations, for example, in the real world.

Reviewer: ```suggestion
0xB10C: 00:22:49 Right, so the goal of `Erlay` is to reduce the bandwidth usage for transaction propagation.And one thing I did with the `tracepoints` is I ran an `Erlay` patch node from the `PR` and I ran master node and compared those two and the bandwidths that we're using, they're connecting to the same peers and we could actually measure the bandwidth usage of both.We saw the Erlay node using far less, I think only 85% or so, or even less bandwidth for `transaction relay` than the master node.So that's really, and I think that was really helpful for Gleb or at least he communicated that he needs people to actually evaluate his changes and backtest his simulations, for example, in the real world.
```

Reviewer: Erlay

> Mark Erhardt: 00:23:34 So that's really interesting.Were the other peers that you were testing against also running the early patch?

Reviewer: ```suggestion
Mark Erhardt: 00:23:34 So that's really interesting.Were the other peers that you were testing against also running the Erlay patch?
```

0xB10C: 00:23:40 Yeah.Yeah.Okay.Yeah.So Gleb ran, I think, 12 Erlay peers.I ran one master and then one Erlay peer.

> Mark Erhardt: 00:23:49 And yeah.So your non-early peer or your node that wasn't running early and the one that was running early were all connecting to early peers.

Reviewer: ```suggestion
Mark Erhardt: 00:23:49 And yeah.So your non-Erlay peer or your node that wasn't running Erlay and the one that was running early were all connecting to early peers.
```

0xB10C: 00:23:57 Correct, yeah.

Mark Erhardt: 00:23:57 And it reduced the bandwidth use by 15% or so.

0xB10C: 00:24:01 Yeah, and some occasions even more, yeah.

Adam Jonas: 00:24:04 Cool. Is there anything else we should cover?

0xB10C: 00:24:06 Not at the moment.OK.Maybe next time.

Adam Jonas: 00:24:08 Cool.Thanks, Timo.It's good to have you back.Thanks for telling us about what you've been up to since you were last here.

0xB10C: 00:24:13 Thanks for having me.

Adam Jonas: 00:24:21 All right.So another conversation in the books.Any takeaways from our conversation with Timo?

Mark Erhardt: 00:24:27 That was fun, short and sweet.

Adam Jonas: 00:24:30 He's up to a lot of good things for the health of the ecosystem.

Mark Erhardt: 00:24:33 Yeah, I need to make another shout out.One of my favorite websites, `transactionfee.info`, where I quote a lot of charts from, is also run by Timo.

Adam Jonas: 00:24:42 Well, you can pay him off with a lot of likes.

Mark Erhardt: 00:24:45 I already bought him breakfast.Seems like a fair trade.

Adam Jonas: 00:24:48 All right, well thank you for joining us and we will have hopefully another episode out shortly.Thanks.

> [Theme song]: 00 :25:03 Have fun here

Reviewer: ```suggestion
```

Reviewer: It sounds like "Ha ha". It's not "Have fun here". Let's remove it.
